File: services/api/src/common/enum/tuition.enum.ts
/**
 * 📘 tuition.enum.ts
 * Enum và Tuple dùng chung trong module học phí.
 */

/**
 * Các toán tử điều kiện hỗ trợ trong hệ thống học phí.
 */
export enum TuitionConditionOperatorEnum {
  Eq = "eq",
  Neq = "neq",
  Gt = "gt",
  Lt = "lt",
  Gte = "gte",
  Lte = "lte",
  In = "in",
  Nin = "nin",
  Contains = "contains",
}

/**
 * Tuple được sinh từ enum, dùng để gắn enum constraint trong Mongoose Schema.
 */
export const TuitionConditionOperatorTuple = Object.values(
  TuitionConditionOperatorEnum
) as [TuitionConditionOperatorEnum, ...TuitionConditionOperatorEnum[]];

/**
 * Các kiểu dữ liệu cho giá trị so sánh.
 */
export enum TuitionConditionValueTypeEnum {
  String = "String",
  Number = "Number",
  Boolean = "Boolean",
  Array = "Array",
  Object = "Object",
}

export const TuitionConditionValueTypeTuple = Object.values(
  TuitionConditionValueTypeEnum
) as [TuitionConditionValueTypeEnum, ...TuitionConditionValueTypeEnum[]];

/**
 * 📘 tuition.enum.ts
 * Enum và Tuple dùng chung cho module học phí (tuition).
 */

/* ============================================================
 * ENUM CHO DISCOUNT POLICY
 * ============================================================
 */

/**
 * Loại giảm giá (theo phần trăm hoặc cố định)
 */
export enum DiscountKindEnum {
  Percentage = "Percentage",
  Fixed = "Fixed",
}

export const DiscountKindTuple = Object.values(DiscountKindEnum) as [
  DiscountKindEnum,
  ...DiscountKindEnum[]
];

/**
 * Mục tiêu áp dụng giảm giá
 * - subtotal: tổng tạm tính
 * - baseFee: học phí cơ bản
 * - specific_fee: chỉ áp dụng cho một phụ phí cụ thể
 */
export enum DiscountTargetEnum {
  Subtotal = "Subtotal",
  BaseFee = "BaseFee",
  SpecificFee = "SpecificFee",
}

export const DiscountTargetTuple = Object.values(DiscountTargetEnum) as [
  DiscountTargetEnum,
  ...DiscountTargetEnum[]
];

/**
 * Phạm vi áp dụng (scope)
 */
export enum DiscountApplicabilityScopeEnum {
  All = "All",
  Class = "Class",
  Grade = "Grade",
  Student = "Student",
}

export const DiscountApplicabilityScopeTuple = Object.values(
  DiscountApplicabilityScopeEnum
) as [DiscountApplicabilityScopeEnum, ...DiscountApplicabilityScopeEnum[]];

/**
 * Chu kỳ áp dụng (once per)
 */
export enum DiscountOncePerEnum {
  Month = "Month",
  Term = "Term",
  None = "None", // dùng thay cho null để đồng nhất kiểu
}

export const DiscountOncePerTuple = Object.values(DiscountOncePerEnum) as [
  DiscountOncePerEnum,
  ...DiscountOncePerEnum[]
];

/**
 * 📘 tuition.enum.ts
 * Enum và Tuple dùng chung cho module học phí (Tuition, ExtraFee, DiscountPolicy, Condition).
 */

/* ============================================================
 * EXTRA FEE ENUMS
 * ============================================================
 */

/**
 * Cách tính phụ phí (Extra Fee Calculation Type)
 */
export enum FeeCalcTypeEnum {
  Fixed = "Fixed",
  PerSession = "PerSession",
  PerMonth = "PerMonth",
  PerYear = "PerYear",
}

export const FeeCalcTypeTuple = Object.values(FeeCalcTypeEnum) as [
  FeeCalcTypeEnum,
  ...FeeCalcTypeEnum[]
];

/**
 * Phạm vi áp dụng phụ phí (Scope)
 */
export enum ExtraFeeApplicabilityScopeEnum {
  All = "All",
  Class = "Class",
  Grade = "Grade",
  Student = "Student",
}

export const ExtraFeeApplicabilityScopeTuple = Object.values(
  ExtraFeeApplicabilityScopeEnum
) as [ExtraFeeApplicabilityScopeEnum, ...ExtraFeeApplicabilityScopeEnum[]];

/**
 * Chu kỳ áp dụng phụ phí (Once Per Period)
 */
export enum ExtraFeeOncePerEnum {
  Month = "Month",
  Term = "Term",
  Year = "Year",
  None = "None",
}

export const ExtraFeeOncePerTuple = Object.values(ExtraFeeOncePerEnum) as [
  ExtraFeeOncePerEnum,
  ...ExtraFeeOncePerEnum[]
];

/**
 * ============================================================
 * ENUM CHO TUITION CHÍNH
 * ============================================================
 */

/**
 * Trạng thái của học phí (tuition status)
 */
export enum TuitionStatusEnum {
  Draft = "Draft",
  Pending = "Pending",
  Paid = "Paid",
  Cancelled = "Cancelled",
  Failed = "Failed",
}

export const TuitionStatusTuple = Object.values(TuitionStatusEnum) as [
  TuitionStatusEnum,
  ...TuitionStatusEnum[]
];

/**
 * Loại tiền tệ (currency)
 * → Có thể mở rộng thêm USD, EUR nếu OmniMer EDU có multi-region
 */
export enum CurrencyEnum {
  VND = "VND",
  USD = "USD",
}

export const CurrencyTuple = Object.values(CurrencyEnum) as [
  CurrencyEnum,
  ...CurrencyEnum[]
];

File: apps/omnimereduapp/lib/core/constants/tuition_enum.dart
/**
 * 📘 tuition_enum.dart
 * Enum dùng trong hệ thống học phí (OmniMer EDU)
 * Mỗi enum đều có:
 * - displayName: tên hiển thị thân thiện
 * - asString: lấy giá trị gốc để lưu DB/API
 * - fromString(): parse từ giá trị string
 */

/// Toán tử điều kiện
enum TuitionConditionOperatorEnum {
  eq("Bằng"),
  neq("Khác"),
  gt("Lớn hơn"),
  lt("Nhỏ hơn"),
  gte("Lớn hơn hoặc bằng"),
  lte("Nhỏ hơn hoặc bằng"),
  rin("Nằm trong"),
  nin("Không nằm trong"),
  contains("Chứa");

  final String displayName;
  const TuitionConditionOperatorEnum(this.displayName);

  String get asString => name;

  static TuitionConditionOperatorEnum fromString(String? value) {
    return TuitionConditionOperatorEnum.values.firstWhere(
      (e) => e.name == value,
      orElse: () => TuitionConditionOperatorEnum.eq,
    );
  }
}

/// Kiểu dữ liệu của giá trị điều kiện
enum TuitionConditionValueTypeEnum {
  string("Chuỗi"),
  number("Số"),
  boolean("Đúng/Sai"),
  array("Mảng"),
  object("Đối tượng");

  final String displayName;
  const TuitionConditionValueTypeEnum(this.displayName);

  String get asString => name;

  static TuitionConditionValueTypeEnum fromString(String? value) {
    return TuitionConditionValueTypeEnum.values.firstWhere(
      (e) => e.name == value,
      orElse: () => TuitionConditionValueTypeEnum.string,
    );
  }
}

/// Loại giảm giá
enum DiscountKindEnum {
  percentage("Theo phần trăm"),
  fixed("Cố định");

  final String displayName;
  const DiscountKindEnum(this.displayName);

  String get asString => name;

  static DiscountKindEnum fromString(String? value) {
    return DiscountKindEnum.values.firstWhere(
      (e) => e.name == value,
      orElse: () => DiscountKindEnum.percentage,
    );
  }
}

/// Mục tiêu áp dụng giảm giá
enum DiscountTargetEnum {
  subtotal("Tổng tạm tính"),
  baseFee("Học phí cơ bản"),
  specific_fee("Phụ phí cụ thể");

  final String displayName;
  const DiscountTargetEnum(this.displayName);

  String get asString => name;

  static DiscountTargetEnum fromString(String? value) {
    return DiscountTargetEnum.values.firstWhere(
      (e) => e.name == value,
      orElse: () => DiscountTargetEnum.subtotal,
    );
  }
}

/// Phạm vi áp dụng giảm giá
enum DiscountApplicabilityScopeEnum {
  All("Tất cả"),
  Class("Lớp học"),
  Grade("Khối lớp"),
  Student("Học sinh");

  final String displayName;
  const DiscountApplicabilityScopeEnum(this.displayName);

  String get asString => name;

  static DiscountApplicabilityScopeEnum fromString(String? value) {
    return DiscountApplicabilityScopeEnum.values.firstWhere(
      (e) => e.name == value,
      orElse: () => DiscountApplicabilityScopeEnum.All,
    );
  }
}

/// Chu kỳ áp dụng giảm giá
enum DiscountOncePerEnum {
  month("Theo tháng"),
  term("Theo học kỳ"),
  none("Không giới hạn");

  final String displayName;
  const DiscountOncePerEnum(this.displayName);

  String get asString => name;

  static DiscountOncePerEnum fromString(String? value) {
    return DiscountOncePerEnum.values.firstWhere(
      (e) => e.name == value,
      orElse: () => DiscountOncePerEnum.none,
    );
  }
}

/// Cách tính phụ phí
enum FeeCalcTypeEnum {
  fixed("Cố định"),
  per_session("Theo buổi"),
  per_month("Theo tháng"),
  formula("Theo công thức");

  final String displayName;
  const FeeCalcTypeEnum(this.displayName);

  String get asString => name;

  static FeeCalcTypeEnum fromString(String? value) {
    return FeeCalcTypeEnum.values.firstWhere(
      (e) => e.name == value,
      orElse: () => FeeCalcTypeEnum.fixed,
    );
  }
}

/// Phạm vi áp dụng phụ phí
enum ExtraFeeApplicabilityScopeEnum {
  All("Tất cả"),
  Class("Lớp học"),
  Grade("Khối lớp"),
  Student("Học sinh");

  final String displayName;
  const ExtraFeeApplicabilityScopeEnum(this.displayName);

  String get asString => name;

  static ExtraFeeApplicabilityScopeEnum fromString(String? value) {
    return ExtraFeeApplicabilityScopeEnum.values.firstWhere(
      (e) => e.name == value,
      orElse: () => ExtraFeeApplicabilityScopeEnum.All,
    );
  }
}

/// Chu kỳ áp dụng phụ phí
enum ExtraFeeOncePerEnum {
  month("Theo tháng"),
  term("Theo học kỳ"),
  year("Theo năm"),
  none("Không giới hạn");

  final String displayName;
  const ExtraFeeOncePerEnum(this.displayName);

  String get asString => name;

  static ExtraFeeOncePerEnum fromString(String? value) {
    return ExtraFeeOncePerEnum.values.firstWhere(
      (e) => e.name == value,
      orElse: () => ExtraFeeOncePerEnum.none,
    );
  }
}

/// Loại công thức tính phụ phí
enum ExtraFeeFormulaTypeEnum {
  jsonlogic("JSON Logic"),
  js("JavaScript");

  final String displayName;
  const ExtraFeeFormulaTypeEnum(this.displayName);

  String get asString => name;

  static ExtraFeeFormulaTypeEnum fromString(String? value) {
    return ExtraFeeFormulaTypeEnum.values.firstWhere(
      (e) => e.name == value,
      orElse: () => ExtraFeeFormulaTypeEnum.jsonlogic,
    );
  }
}

/// Trạng thái học phí
enum TuitionStatusEnum {
  draft("Nháp"),
  pending("Chờ thanh toán"),
  paid("Đã thanh toán"),
  cancelled("Đã hủy"),
  failed("Thất bại");

  final String displayName;
  const TuitionStatusEnum(this.displayName);

  String get asString => name;

  static TuitionStatusEnum fromString(String? value) {
    return TuitionStatusEnum.values.firstWhere(
      (e) => e.name == value,
      orElse: () => TuitionStatusEnum.draft,
    );
  }
}

/// Loại tiền tệ
enum CurrencyEnum {
  vnd("Việt Nam Đồng"),
  usd("Đô la Mỹ");

  final String displayName;
  const CurrencyEnum(this.displayName);

  String get asString => name.toUpperCase();

  static CurrencyEnum fromString(String? value) {
    return CurrencyEnum.values.firstWhere(
      (e) => e.name.toUpperCase() == value?.toUpperCase(),
      orElse: () => CurrencyEnum.vnd,
    );
  }
}
